language-server/src/dotty/tools/languageserver/DottyLanguageServer.scala

```diff
@@ -0,0 +1,438 @@
+  }
+
+  /** The driver instance responsible for compiling `uri` */
+  def driverFor(uri: URI): InteractiveDriver = {
+    val matchingConfig =
+      drivers.keys.find(config => config.sourceDirectories.exists(sourceDir =>
+        uri.getRawPath.startsWith(sourceDir.getAbsolutePath.toString)))
+    matchingConfig match {
+      case Some(config) =>
+        drivers(config)
+      case None =>
+        val config = drivers.keys.head
+        println(s"No configuration contains $uri as a source file, arbitrarily choosing ${config.id}")
+        drivers(config)
+    }
+  }
+
+  override def connect(client: LanguageClient): Unit = {
+    this.client = client
+  }
+
+  override def exit(): Unit = {
+    System.exit(0)
+  }
+
+  override def shutdown(): CompletableFuture[Object] = {
+    CompletableFuture.completedFuture(new Object)
+  }
+
+  private[this] def computeAsync[R](fun: CancelChecker => R): CompletableFuture[R] =
+    CompletableFutures.computeAsync({(cancelToken: CancelChecker) =>
+      // We do not support any concurrent use of the compiler currently.
+      thisServer.synchronized {
+        cancelToken.checkCanceled()
+        try {
+          fun(cancelToken)
+        } catch {
+          case NonFatal(ex) =>
+            ex.printStackTrace
+            throw ex
+        }
+      }
+    })
+
+  override def initialize(params: InitializeParams) = computeAsync { cancelToken =>
+    rootUri = params.getRootUri
+    assert(rootUri != null)
+
+    val c = new ServerCapabilities
+    c.setTextDocumentSync(TextDocumentSyncKind.Full)
+    c.setDocumentHighlightProvider(true)
+    c.setDocumentSymbolProvider(true)
+    c.setDefinitionProvider(true)
+    c.setRenameProvider(true)
+    c.setHoverProvider(true)
+    c.setWorkspaceSymbolProvider(true)
+    c.setReferencesProvider(true)
+    c.setCompletionProvider(new CompletionOptions(
+      /* resolveProvider = */ false,
+      /* triggerCharacters = */ List(".").asJava))
+
+    // Do most of the initialization asynchronously so that we can return early
+    // from this method and thus let the client know our capabilities.
+    CompletableFuture.supplyAsync(() => drivers)
+
+    new InitializeResult(c)
+  }
+
+  override def didOpen(params: DidOpenTextDocumentParams): Unit = thisServer.synchronized {
+    val document = params.getTextDocument
+    val uri = new URI(document.getUri)
+    val driver = driverFor(uri)
+
+    val text = document.getText
+    val diags = driver.run(uri, text)
+
+    client.publishDiagnostics(new PublishDiagnosticsParams(
+      document.getUri,
+      diags.flatMap(diagnostic).asJava))
+  }
+
+  override def didChange(params: DidChangeTextDocumentParams): Unit = thisServer.synchronized {
+    val document = params.getTextDocument
+    val uri = new URI(document.getUri)
+    val driver = driverFor(uri)
+
+    val change = params.getContentChanges.get(0)
+    assert(change.getRange == null, "TextDocumentSyncKind.Incremental support is not implemented")
```

Comment:
Why is this an assert? Is it to show an incorrectly set up client an error? There isn't something like "client error" that can be sent back instead?